walletnode/services/userdataprocess/task.go

```diff
@@ -0,0 +1,391 @@
+	topNodes.DisconnectInactive()
+
+	// Cancel context when any connection is broken.
+	groupConnClose, _ := errgroup.WithContext(ctx)
+	groupConnClose.Go(func() error {
+		defer cancel()
+		return nodes.WaitConnClose(ctx)
+	})
+	task.UpdateStatus(StatusConnected)
+	task.nodes = nodes
+
+	if task.request == nil {
+		// PROCESS TO RETRIEVE USERDATA FROM METADATA LAYER
+		if err := nodes.ReceiveUserdata(ctx, task.userpastelid); err != nil {
+			return errors.Errorf("failed to receive userdata: %w", err)
+		}
+		// Post on result channel
+		node := nodes[0]
+		if node.ResultGet != nil {
+			task.resultChanGet <- node.ResultGet
+		} else {
+			return errors.Errorf("failed to receive userdata")
+		}
+
+		log.WithContext(ctx).Debug("Finished retrieve userdata")
+
+	} else {
+		// PROCESS TO SET/UPDATE USERDATA TO METADATA LAYER
+		// Get the previous block hash
+		// Get block num
+		blockHash := ""
+		blockNum, err := task.pastelClient.GetBlockCount(ctx)
+		if err != nil {
+			log.WithContext(ctx).Debug("failed to get block num: %w", err)
+		} else {
+			// Get block hash string
+			blockInfo, err := task.pastelClient.GetBlockVerbose1(ctx, blockNum)
+			if err != nil {
+				log.WithContext(ctx).Debug("failed to get block info with error: %w", err)
+			} else {
+				blockHash = blockInfo.Hash
+			}
+		}
+
+		task.request.PreviousBlockHash = blockHash
+
+		// Get the value of task.request.ArtistPastelIDPassphrase for sign data, then empty it in the request to make sure it not sent to supernodes
+		passphrase := task.request.ArtistPastelIDPassphrase
+		task.request.ArtistPastelIDPassphrase = ""
+		// Marshal task.request to byte array for signing
+		js, err := json.Marshal(task.request)
+		if err != nil {
+			return errors.Errorf("failed to encode request %w", err)
+		}
+
+		// Hash the request
+		hashvalue, err := userdata.Sha3256hash(js)
+		if err != nil {
+			return errors.Errorf("failed to hash request %w", err)
+		}
+
+		// Sign request with Wallet Node's pastelID and passphrase
+		signature, err := task.pastelClient.Sign(ctx, hashvalue, task.request.ArtistPastelID, passphrase, "ed448")
+		if err != nil {
+			return errors.Errorf("failed to sign ticket %w", err)
+		}
+
+		userdata := &userdata.ProcessRequestSigned{
+			Userdata:     task.request,
+			UserdataHash: hex.EncodeToString(hashvalue),
+			Signature:    hex.EncodeToString(signature),
+		}
+
+		// Send userdata to supernodes for storing in MDL's rqlite db.
+		if err := nodes.SendUserdata(ctx, userdata); err != nil {
+			return err
+		}
+		res, err := task.AggregateResult(ctx, nodes)
+		if err != nil {
+			return err
+		}
+		// Post on result channel
+		task.resultChan <- &res
+		log.WithContext(ctx).WithField("userdata_result", res).Debug("Posted userdata result")
+	}
+
+	// close the connections
+	for i := range nodes {
+		if err := nodes[i].Connection.Close(); err != nil {
+			return errors.Errorf("failed to close connection to node %s %w", nodes[i].PastelID(), err)
```

Comment:
Will return here cause the reset of connections cannot be closed ?

Reply:
just print error log instead of returning error here
thank you